pulumi import -f panics if the file is missing an ID

### What happened?

I ran `pulumi import -f import.json` with a file containing roughly:

```json
{
	"resources": [
		{
			"type": "someTokenType",
			"name": "someName"
		}
	]
}
```

This panicked with:

```
panic: fatal: An assertion has failed

                                     goroutine 66 [running]:
                                                            github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.failfast(...)
                                                                                                                                        /home/runner/work/pulumi/pulumi/sdk/go/common/util/contract/failfast.go:23
                github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.Assert(...)
                                                                                        /home/runner/work/pulumi/pulumi/sdk/go/common/util/contract/assert.go:26
                                                                                                                                                                github.com/pulumi/pulumi/pkg/v3/resource/deploy.newImportDeploymentStep(0x0?, 0xc001571d10?, {0xc0015ab900?, 0x1?, 0x0?})
                                                                                        /home/runner/work/pulumi/pulumi/pkg/resource/deploy/step.go:886 +0x319
                                                                                                                                                              github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*importer).importResources(0xc000f95948, {0x2b0a3d0, 0xc0015a3b80})
                                                                                        /home/runner/work/pulumi/pulumi/pkg/resource/deploy/import.go:359 +0x805
                                                                                                                                                                github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*deploymentExecutor).importResources(0xc0015a1e18, {0x2b0a478, 0xc00163d860}, {{0x79ed181aadb0, 0xc0003d2900}, 0x7fffffff, 0x0, 0x0, {{0x0, 0x0, ...}, ...}, ...}, ...)
                                                                                                                                                                                        /home/runner/work/pulumi/pulumi/pkg/resource/deploy/deployment_executor.go:455 +0x132
                                                                           github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*deploymentExecutor).Execute(0xc0015a1e18, {0x2b0a478?, 0xc00163d860}, {{0x79ed181aadb0, 0xc0003d2900}, 0x7fffffff, 0x0, 0x0, {{0x0, 0x0, ...}, ...}, ...}, ...)
                                                                                                /home/runner/work/pulumi/pulumi/pkg/resource/deploy/deployment_executor.go:144 +0xe38
                                                                                                                                                                                     github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*Deployment).Execute(0xc001650160, {0x2b0a478, 0xc00163d860}, {{0x79ed181aadb0, 0xc0003d2900}, 0x7fffffff, 0x0, 0x0, {{0x0, 0x0, ...}, ...}, ...}, ...)
                                                                                                                                                                                                /home/runner/work/pulumi/pulumi/pkg/resource/deploy/deployment.go:534 +0xd0
                                                                         github.com/pulumi/pulumi/pkg/v3/engine.(*deployment).run.func1()
                                                                                                                                                /home/runner/work/pulumi/pulumi/pkg/engine/deployment.go:291 +0x290
                 created by github.com/pulumi/pulumi/pkg/v3/engine.(*deployment).run
                                                                                        /home/runner/work/pulumi/pulumi/pkg/engine/deployment.go:274 +0x365
```

Note that the formatting is all over the place because the panic also left the terminal in a bad mode.

### Expected Behavior

`pulumi import` succeeds or gives me a meaningful error message.

Instead, it only prints "panic: fatal: An assertion has failed".

### Steps to reproduce

- Create a new project

- Create a file import.json with the contents:

    ```json
    {
        "resources": [
            {
                "type": "someTokenType",
                "name": "someName"
            }
        ]
    }
    ```

- Run `pulumi import -f import.json`

### Output of `pulumi about`

```
% pulumi about
CLI
Version      3.54.0
Go Version   go1.20
Go Compiler  gc

Plugins
NAME    VERSION
github  4.17.0
go      unknown

Host
OS       arch
Version
Arch     x86_64

This project is written in go: executable='/usr/bin/go' version='go version go1.20 linux/amd64'

[..]


Dependencies:
NAME                                    VERSION
github.com/pulumi/pulumi-github/sdk/v4  4.17.0
github.com/pulumi/pulumi/sdk/v3         3.51.0
```

### Additional context

The error is coming from this line. https://github.com/pulumi/pulumi/blob/8915911825ba26aea1ad4a53aa666eacff8ca7da/pkg/resource/deploy/step.go#L886

The file I supplied was obviously invalid, but that should not result in an inscrutable panic.

Perhaps as a general practice, we should replace all contract.Asserts with Assertfs so that we're forced to supply a message. (And, where possible, we should just use errors instead of contract.Assert.)

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

--- a/changelog/pending/20230215--cli-import--fixes-panic-on-incomplete-resources-in-json-file.yaml
+++ b/changelog/pending/20230215--cli-import--fixes-panic-on-incomplete-resources-in-json-file.yaml
@@ -1,0 +1,4 @@
+changes:
+- type: fix
+  scope: cli/import
+  description: Fixes panic on incomplete resources in JSON file.

--- a/pkg/cmd/pulumi/import.go
+++ b/pkg/cmd/pulumi/import.go
@@ -157,8 +157,46 @@ func parseImportFile(f importFile, protectResources bool) ([]deploy.Import, impo
 		names[urn] = name
 	}
 
+	// Attempts to generate a human-readable description of the given import spec
+	// for use in error messages using whatever information is available.
+	// For example:
+	//
+	//	resource 'foo' of type 'aws:ec2/vpc:Vpc'
+	//	resource 'foo'
+	//	resource 3 of type 'aws:ec2/vpc:Vpc'
+	//	resource 3
+	describeResource := func(idx int, spec importSpec) string {
+		var sb strings.Builder
+		sb.WriteString("resource ")
+
+		switch {
+		case spec.Name != "":
+			fmt.Fprintf(&sb, "'%v'", spec.Name)
+		case spec.ID != "":
+			fmt.Fprintf(&sb, "'%v'", spec.ID)
+		default:
+			fmt.Fprintf(&sb, "%d", idx)
+		}
+
+		if spec.Type != "" {
+			fmt.Fprintf(&sb, " of type '%v'", spec.Type)
+		}
+
+		return sb.String()
+	}
+
 	imports := make([]deploy.Import, len(f.Resources))
 	for i, spec := range f.Resources {
+		if spec.Type == "" {
+			return nil, nil, fmt.Errorf("%v has no type", describeResource(i, spec))
+		}
+		if spec.Name == "" {
+			return nil, nil, fmt.Errorf("%v has no name", describeResource(i, spec))
+		}
+		if spec.ID == "" {
+			return nil, nil, fmt.Errorf("%v has no ID", describeResource(i, spec))
+		}
+
 		imp := deploy.Import{
 			Type:       spec.Type,
 			Name:       spec.Name,
@@ -170,8 +208,8 @@ func parseImportFile(f importFile, protectResources bool) ([]deploy.Import, impo
 		if spec.Parent != "" {
 			urn, ok := f.NameTable[spec.Parent]
 			if !ok {
-				return nil, nil, fmt.Errorf("the parent '%v' for resource '%v' of type '%v' has no name",
-					spec.Parent, spec.Name, spec.Type)
+				return nil, nil, fmt.Errorf("the parent '%v' for %v has no name",
+					spec.Parent, describeResource(i, spec))
 			}
 			imp.Parent = urn
 		}
@@ -179,8 +217,8 @@ func parseImportFile(f importFile, protectResources bool) ([]deploy.Import, impo
 		if spec.Provider != "" {
 			urn, ok := f.NameTable[spec.Provider]
 			if !ok {
-				return nil, nil, fmt.Errorf("the provider '%v' for resource '%v' of type '%v' has no name",
-					spec.Provider, spec.Name, spec.Type)
+				return nil, nil, fmt.Errorf("the provider '%v' for %v has no name",
+					spec.Provider, describeResource(i, spec))
 			}
 			imp.Provider = urn
 		}
@@ -188,8 +226,8 @@ func parseImportFile(f importFile, protectResources bool) ([]deploy.Import, impo
 		if spec.Version != "" {
 			v, err := semver.ParseTolerant(spec.Version)
 			if err != nil {
-				return nil, nil, fmt.Errorf("could not parse version '%v' for resource '%v' of type '%v': %w",
-					spec.Version, spec.Name, spec.Type, err)
+				return nil, nil, fmt.Errorf("could not parse version '%v' for %v: %w",
+					spec.Version, describeResource(i, spec), err)
 			}
 			imp.Version = &v
 		}

--- a/pkg/cmd/pulumi/import_test.go
+++ b/pkg/cmd/pulumi/import_test.go
@@ -1,0 +1,114 @@
+package main
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestParseImportFile_errors(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		desc    string
+		give    importFile
+		wantErr string
+	}{
+		{
+			desc:    "missing everything",
+			give:    importFile{Resources: []importSpec{{}}},
+			wantErr: "resource 0 has no type",
+		},
+		{
+			desc: "missing type",
+			give: importFile{
+				Resources: []importSpec{
+					{
+						Name: "foo",
+						ID:   "bar",
+					},
+				},
+			},
+			wantErr: "resource 'foo' has no type",
+		},
+		{
+			desc: "missing name",
+			give: importFile{
+				Resources: []importSpec{
+					{
+						ID:   "bar",
+						Type: "foo:bar:baz",
+					},
+				},
+			},
+			wantErr: "resource 'bar' of type 'foo:bar:baz' has no name",
+		},
+		{
+			desc: "missing id",
+			give: importFile{
+				Resources: []importSpec{
+					{
+						Name: "foo",
+						Type: "foo:bar:baz",
+					},
+				},
+			},
+			wantErr: "resource 'foo' of type 'foo:bar:baz' has no ID",
+		},
+		{
+			desc: "missing parent",
+			give: importFile{
+				Resources: []importSpec{
+					{
+						Name:   "thing",
+						ID:     "thing",
+						Parent: "unknown",
+						Type:   "foo:bar:baz",
+					},
+				},
+			},
+			wantErr: "the parent 'unknown' for resource 'thing' of type 'foo:bar:baz' has no name",
+		},
+		{
+			desc: "missing provider",
+			give: importFile{
+				Resources: []importSpec{
+					{
+						Name:     "thing",
+						ID:       "thing",
+						Provider: "unknown",
+						Type:     "foo:bar:baz",
+					},
+				},
+			},
+			wantErr: "the provider 'unknown' for resource 'thing' of type 'foo:bar:baz' has no name",
+		},
+		{
+			desc: "bad version",
+			give: importFile{
+				Resources: []importSpec{
+					{
+						Name:    "thing",
+						ID:      "thing",
+						Type:    "foo:bar:baz",
+						Version: "not-a-semver",
+					},
+				},
+			},
+			wantErr: "could not parse version 'not-a-semver' for resource 'thing' of type 'foo:bar:baz'",
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.desc, func(t *testing.T) {
+			t.Parallel()
+
+			require.NotEmpty(t, tt.wantErr, "invalid test: wantErr must not be empty")
+
+			_, _, err := parseImportFile(tt.give, false)
+			assert.ErrorContains(t, err, tt.wantErr)
+		})
+	}
+}

